sdk/src/main/scala/com/horizen/network/SyncStatusActor.scala

```diff
@@ -0,0 +1,275 @@
+  }
+
+  // return true if the difference between current timestamp and applied block timestamp is less than block rate * confidence parameter
+  // return false otherwise
+  def checkAppliedBlockTimestamp(blockTimestamp: Long): Boolean = {
+    (timeProvider.time() - blockTimestamp).compareTo(params.consensusSecondsInSlot * confidenceParameter) < 0
+  }
+
+  // In this method the message SemanticallySuccessfulModifier is processed
+  // this message is sent when a new block is applied to the state
+  // the new blocks are processed in the sync status actor if the processNewBlocks is active
+  // if the syncStatus we retrieve the current block from the view and calculate the potential highest block (given the current timestamp)
+  // otherwise the currentBlock counter is updated
+  // if the syncStatus is true and the applied block timestamp is near the current timestamp (block rate * confidence parameter)
+  // we set the node as synched and reset the internal state
+  def processNewBlockApplied(sidechainBlock: PMOD): Unit = {
+
+    if(processNewBlocks) {
+
+      // if the node is not (yet) syncing we process the new block and update the syncStatus
+      if (!syncStatus) {
+
+        // retrieve the current block height from the node view
+        val currentHeightFromView = Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView(retrieveCurrentHeight), timeoutDuration)
+          .asInstanceOf[Int]
+
+        // calculate the block correction needed to calculate the potential highest block
+        val blockCorrection = Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView((view: View) =>
+          calculateBlockCorrection(view, currentHeightFromView)), timeoutDuration)
+          .asInstanceOf[Double]
+
+        // calculate the potential highest block
+        val currentTime: Long = timeProvider.time() / 1000
+        val timestampDifference = currentTime - sidechainBlock.timestamp
+        val blocksFromCurrentToTip = ((timestampDifference / params.consensusSecondsInSlot.toDouble) * blockCorrection).toInt
+
+        // set the current internal state
+        syncStatus = true
+        currentBlock = currentHeightFromView
+        startingBlock = currentHeightFromView
+        highestBlock = currentHeightFromView + blocksFromCurrentToTip
+
+        // broadcast sync start and its details, starting block is equal to current block in this case
+        val syncStatusMessage = new SyncStatus(true, BigInt(currentBlock), BigInt(startingBlock), BigInt(highestBlock))
+        context.system.eventStream.publish(NotifySyncStart(syncStatusMessage))
+
+      }
+      // if the node is already syncing we update the current block value adding one for every new block that is applied on the state
+      // we also check if the applied block timestamp is near the current timestamp, in this case we set the syncStatus to false and reset the state
+      else {
+        // update the current block counter
+        currentBlock += 1
```

Comment:
time to time we need to actualize the prediction of highest block.
At the moment the algorithm works as:
1) start sync from scratch -> height is `< previousBlocksRange` -> correction = `1` -> we predict the tip height that will be for sure much higher then real one
2) start when we have height `>= previousBlocksRange` -> correction = `N >= 1` -> we predict the tip to some value, and never update. Let's assume that the first 100 blocks were ver-very rare (1 out of 50 slots occupied) -> prediction is very underestimated. User will see that he is syncing but the `currentBlock > `highestBlock`.

Use average mean or even weighted average mean.

Reply:
The method to calculate the estimated highest block was moved to a separate class (`SyncStatusUtil`), here I defined a class and the related companion object to be able to use the methods as static.
The block correction is calculated using weighted average mean between:
- the block correction calculated from the genesis block with a weight equal to 0.5
- the block correction calculated from the sidechain half with a weight equal to 1.0
In this way we are able to calculate a potential highest block giving more weight to the most recent behavior. In case we are planning to change the block rate radically in the future maybe it's better to use a different calculation logic not using the older data.

The calculation is performed if the current height is more than a threshold value set at 15000 (approximately equal to 3/4 days of sidechain activity with 12s block time and ~0.65 block correction), otherwise a default value (0.65 is returned). The default value was calculated on the current testnet network.

The estimated highest block value is re-calculated every 20000 blocks or if the current block become equal to this value.

If needed we can modify or discuss these values used